File: Dsa-day51.java
class Solution {
    public long minOperations(int[][] queries) {
        long[] pow4 = new long[20];
        pow4[0] = 1;
        int len = 1;
        while (pow4[len - 1] <= 1_000_000_000L) {
            pow4[len] = pow4[len - 1] * 4;
            len++;
        }

        long ans = 0;

        for (int[] q : queries) {
            long l = q[0], r = q[1];
            long totalOps = 0;

            for (int k = 1; k < len; k++) {
                long left = Math.max(l, pow4[k - 1]);
                long right = Math.min(r, pow4[k] - 1);
                if (left <= right) {
                    long cnt = right - left + 1;
                    totalOps += cnt * k;
                }
            }

            ans += (totalOps + 1) / 2; 
        }

        return ans;
    }
}

File: Dsa-day51.cpp
class Solution {
public:
    long long minOperations(vector<vector<int>>& queries) {
        vector<long long> pow4 = {1};
        while (pow4.back() <= 1e9) {
            pow4.push_back(pow4.back() * 4);
        }
        
        long long ans = 0;
        
        for (auto &q : queries) {
            long long l = q[0], r = q[1];
            long long totalOps = 0;
            
            for (int k = 1; k < (int)pow4.size(); k++) {
                long long left = max(l, pow4[k-1]);
                long long right = min(r, pow4[k]-1);
                if (left <= right) {
                    long long cnt = right - left + 1;
                    totalOps += cnt * k;
                }
            }
            
            ans += (totalOps + 1) / 2;  
        }
        
        return ans;
    }
};
